src/xcode/ENA/ENA/Source/Services/Risk/Provider/Revocation/RevocationProvider.swift

```diff
@@ -0,0 +1,266 @@
+//
+// 🦠 Corona-Warn-App
+//
+
+import Foundation
+import SwiftUI
+
+protocol RevocationProviding {
+	func updateCache(with certificates: [HealthCertificate], completion: @escaping () -> Void)
+}
+
+final class RevocationProvider: RevocationProviding {
+
+	// MARK: - Init
+
+	init(
+		_ restService: RestServiceProviding,
+		signatureVerifier: SignatureVerification = SignatureVerifier()
+	) {
+		self.restService = restService
+		self.signatureVerifier = signatureVerifier
+	}
+
+	// MARK: - Protocol RevocationProviding
+
+	func updateCache(with certificates: [HealthCertificate], completion: @escaping () -> Void) {
+		// 1. Filter by certificate type
+		let filteredCertificates = certificates.filter { certificate in
+			(certificate.type == .vaccination ||
+			 certificate.type == .recovery)
+		}
+
+		// 2. group certificates by kid
+		let groupedCertificates = Dictionary(grouping: filteredCertificates) { element in
+			Data(base64Encoded: element.keyIdentifier)?.toHexString() ?? ""
+		}
+
+		// 3. Update KID List
+		let resource = KIDListResource(signatureVerifier: signatureVerifier)
+		restService.load(resource) { [weak self] result in
+			guard let self = self else {
+				Log.error("Failed to get strong self")
+				return
+			}
+
+			switch result {
+			case .failure(let error):
+				Log.error("failed to update kid list", error: error)
+				completion()
+			case .success(let kidList):
+				// helping step -> convert to KidWithTypes model to make things a bit easier
+				let keyIdentifiersWithTypes: [KidWithTypes] = kidList.items.map {
+					let keyIdentifier = $0.kid.toHexString()
+					let types = $0.hashTypes.map { $0.toHexString() }
+					return KidWithTypes(
+						kid: keyIdentifier,
+						types: types
+					)
+				}
+
+				// 4. filter certificates for kid
+				let kids = keyIdentifiersWithTypes.map { $0.kid }
+				let filteredGroupedCertificates = groupedCertificates
+					.filter { !$0.key.isEmpty }
+					.filter { kids.contains($0.key) }
+
+				// 5. calculate revocation coordinates based on kid list for ever filteredGroupedCertificates
+				// incl. 6. Group by RLC by KID Type
+				var certificateByRLC: [RevocationLocation] = []
+				for kidWithTypes in keyIdentifiersWithTypes {
+					let kid = kidWithTypes.kid
+					guard let certificates = filteredGroupedCertificates[kid] else {
+						Log.info("no matching certificates found, nothing to do for that kid")
+						continue
+					}
+
+					let kidTypes = kidWithTypes.types
+					for certificate in certificates {
+						for type in kidTypes {
+							guard let hash = certificate.hash(by: type) else {
+								Log.error("Missing hash value")
+								continue
+							}
+							certificateByRLC.insert(
+								coordinate: RevocationCoordinate(hash: hash),
+								kid: kid,
+								type: type,
+								certificate: certificate
+							)
+						}
+					}
+				}
+				certificateByRLC.sort { lhs, rhs in
+					lhs.type < rhs.type
+				}
+
+				DispatchQueue.global().async {
```

Comment:
Wondering why this dispatch is needed? Aren't we on a background thread already at this point?

Reply:
We moved the dispatch inside the function now and added a comment to explain that this is just a safety measure to ensure that the dispatch group will not block the main thread, regardless of the thread on which it is called.